File: DXDAL/AuditableBase.vb
Imports DevExpress.Xpo
Imports DXDAL.SPGData

<NonPersistent()>
Public Class AuditableBase
    Inherits XPObject

    Protected Sub New()

    End Sub

    Protected Sub New(ByVal session As Session)
        MyBase.New(session)

    End Sub

    Protected Sub New(ByVal session As Session, ByVal classInfo As DevExpress.Xpo.Metadata.XPClassInfo)
        MyBase.New(session, classInfo)

    End Sub

    Private changes As List(Of Change) = New List(Of Change)

    Private Structure Change

        Public PropertyName As String
        Public PrevValue As String
        Public Value As String

    End Structure

    Public ReadOnly Property HasChanges() As Boolean
        Get
            Return changes.Count > 0
        End Get

    End Property

    Private Sub UpdateAudit(ByVal recordID As Integer)

        If changes.Count > 0 Then
            Dim audit = New AuditTrail(Session) With {.ChangeDate = Now, .RecordType = ClassInfo.ClassType.Name, .UserID = Environment.UserName, .RecordID = recordID}

            For Each change In changes
                audit.RecordChanges &= String.Format("{0}:{1}({2});", change.PropertyName, change.PrevValue, change.Value)
            Next
            audit.Save()
        End If

    End Sub

    Protected Overrides Sub OnSaved()
        MyBase.OnSaved()

        UpdateAudit(CInt(GetMemberValue("Oid")))
    End Sub

    Protected Overrides Sub OnChanged(ByVal propertyName As String, ByVal oldValue As Object, ByVal newValue As Object)
        MyBase.OnChanged(propertyName, oldValue, newValue)

        Dim change = New Change() With {.PropertyName = propertyName}
        If oldValue Is Nothing Then
            change.PrevValue = "NULL"
        Else
            change.PrevValue = oldValue.ToString
        End If
        If newValue Is Nothing Then
            change.Value = "NULL"
        Else
            change.Value = newValue.ToString
        End If
        If Not (change.PropertyName = "Oid" AndAlso change.PrevValue = "-1") AndAlso change.PropertyName <> "OptimisticLockField" AndAlso change.PropertyName <> "OptimisticLockFieldInDataLayer" AndAlso change.PropertyName <> "GCRecord" Then
            changes.Add(change)
        End If


    End Sub

    Protected Overrides Sub OnDeleted()
        MyBase.OnDeleted()

        Dim recordID As Integer = CInt(GetPropertyValue("Oid"))

        Dim audit = New AuditTrail(Session) With {.ChangeDate = Now, .RecordType = ClassInfo.ClassType.Name, .UserID = Environment.UserName, .RecordID = recordID, .RecordChanges = recordID.ToString & " was deleted"}

        audit.Save()

    End Sub

End Class

File: DXDAL/AuditableBase.cs
using System.Collections.Generic;
using System;
using DevExpress.Xpo;
using DXDAL.SPGData;

namespace DXDAL
{
    [NonPersistent()]
    public class AuditableBase : XPObject
    {
        protected AuditableBase()
        {
        }

        protected AuditableBase(Session session) : base(session)
        {
        }

        protected AuditableBase(Session session, DevExpress.Xpo.Metadata.XPClassInfo classInfo) : base(session, classInfo)
        {
        }

        private List<Change> changes = new List<Change>();

        private struct Change
        {
            public string PropertyName;
            public string PrevValue;
            public string Value;
        }

        public bool HasChanges => changes.Count > 0;

        private void UpdateAudit(int recordID)
        {
            if (changes.Count > 0)
            {
                AuditTrail audit = new AuditTrail(Session) { ChangeDate =DateTime.Now, RecordType = ClassInfo.ClassType.Name, UserID = Environment.UserName, RecordID = recordID };

                foreach (Change change in changes)
                    audit.RecordChanges += string.Format("{0}:{1}({2});", change.PropertyName, change.PrevValue, change.Value);
                audit.Save();
            }
        }

        protected override void OnSaved()
        {
            base.OnSaved();

            UpdateAudit(System.Convert.ToInt32(GetMemberValue("Oid")));
        }

        protected override void OnChanged(string propertyName, object oldValue, object newValue)
        {
            base.OnChanged(propertyName, oldValue, newValue);

            Change change = new Change() { PropertyName = propertyName };
            if (oldValue == null)
                change.PrevValue = "NULL";
            else
                change.PrevValue = oldValue.ToString();
            if (newValue == null)
                change.Value = "NULL";
            else
                change.Value = newValue.ToString();
            if (!(change.PropertyName == "Oid" && change.PrevValue == "-1") && change.PropertyName != "OptimisticLockField" && change.PropertyName != "OptimisticLockFieldInDataLayer" && change.PropertyName != "GCRecord")
                changes.Add(change);
        }

        protected override void OnDeleted()
        {
            base.OnDeleted();

            int recordID = System.Convert.ToInt32(GetPropertyValue("Oid"));

            AuditTrail audit = new AuditTrail(Session) { ChangeDate = DateTime.Now, RecordType = ClassInfo.ClassType.Name, UserID = Environment.UserName, RecordID = recordID, RecordChanges = recordID.ToString() + " was deleted" };

            audit.Save();
        }
    }
}
